pkg/apiserver/rest/dualwriter_mode1_test.go

```diff
@@ -2,46 +2,237 @@ package rest
 
 import (
 	"context"
+	"errors"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
 	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/apis/example"
 )
 
 const kind = "dummy"
 
-func TestMode1(t *testing.T) {
-	var ls = (LegacyStorage)(nil)
-	var s = (Storage)(nil)
-	lsSpy := NewLegacyStorageSpyClient(ls)
-	sSpy := NewStorageSpyClient(s)
-
-	dw := NewDualWriterMode1(lsSpy, sSpy)
-
-	// Create: it should use the Legacy Create implementation
-	_, err := dw.Create(context.Background(), &dummyObject{}, func(context.Context, runtime.Object) error { return nil }, &metav1.CreateOptions{})
-	assert.NoError(t, err)
-	assert.Equal(t, 1, lsSpy.Counts("LegacyStorage.Create"))
-	assert.Equal(t, 0, sSpy.Counts("Storage.Create"))
-
-	// Get: it should use the Legacy Get implementation
-	_, err = dw.Get(context.Background(), kind, &metav1.GetOptions{})
-	assert.NoError(t, err)
-	assert.Equal(t, 1, lsSpy.Counts("LegacyStorage.Get"))
-	assert.Equal(t, 0, sSpy.Counts("Storage.Get"))
-
-	// List: it should use the Legacy List implementation
-	_, err = dw.List(context.Background(), &metainternalversion.ListOptions{})
-	assert.NoError(t, err)
-	assert.Equal(t, 1, lsSpy.Counts("LegacyStorage.List"))
-	assert.Equal(t, 0, sSpy.Counts("Storage.List"))
-
-	// Delete: it should use the Legacy Delete implementation
-	var deleteValidation = func(ctx context.Context, obj runtime.Object) error { return nil }
-	_, _, err = dw.Delete(context.Background(), kind, deleteValidation, &metav1.DeleteOptions{})
-	assert.NoError(t, err)
-	assert.Equal(t, 1, lsSpy.Counts("LegacyStorage.Delete"))
-	assert.Equal(t, 0, sSpy.Counts("Storage.Delete"))
+var exampleObj = &example.Pod{TypeMeta: metav1.TypeMeta{}, ObjectMeta: metav1.ObjectMeta{Name: "foo", ResourceVersion: "1"}, Spec: example.PodSpec{}, Status: example.PodStatus{}}
+var anotherObj = &example.Pod{TypeMeta: metav1.TypeMeta{}, ObjectMeta: metav1.ObjectMeta{Name: "bar", ResourceVersion: "2"}, Spec: example.PodSpec{}, Status: example.PodStatus{}}
+var failingObj = &example.Pod{TypeMeta: metav1.TypeMeta{}, ObjectMeta: metav1.ObjectMeta{Name: "object-fail", ResourceVersion: "2"}, Spec: example.PodSpec{}, Status: example.PodStatus{}}
+
+func TestMode1_Create(t *testing.T) {
+	type testCase struct {
```

Comment:
Would it be worth combining the test for Create across all modes? We could add additional fields to the test case to indicate whether we expect a Storage/LegacyStorage method to be called. 

Still trying to figure out myself whether it would be worth doing or not. The main reason I'm bringing it up is that it could serve as a helpful record to have all the different create behaviors in one place. What do you think?

Reply:
I guess we could do that, but I'd rather do that after all tests for all modes are in. It's easier to _just_ have to move things :)

Reply:
So true. Yes, let's do that.